src/Lean/Meta/LazyDiscrTree.lean

```diff
@@ -0,0 +1,866 @@
+lazy initialization strategy.
+
+The discrimination tree can be created through
+`createImportedEnvironment`. This creates a discrimination tree from all
+imported modules in an environment using a callback that provides the
+entries as `InitEntry` values.
+
+The function `getMatch` can be used to get the values that match the
+expression as well as an updated lazy discrimination tree that has
+elaborated additional parts of the tree.
+-/
+namespace Lean.Meta.LazyDiscrTree
+
+-- This namespace contains definitions copied from Lean.Meta.DiscrTree.
+namespace MatchClone
+
+/--
+Discrimination tree key.
+-/
+private inductive Key where
+    /-- Constant -/
+  | const : Name → Nat → Key
+  | fvar  : FVarId → Nat → Key
+  | lit   : Literal → Key
+  | star  : Key
+  | other : Key
+  | arrow : Key
+  | proj  : Name → Nat → Nat → Key
+  deriving Inhabited, BEq, Repr
+
+namespace Key
+
+/-- Hash function -/
+protected def hash : Key → UInt64
+  | .const n a   => mixHash 5237 $ mixHash n.hash (hash a)
+  | .fvar n a    => mixHash 3541 $ mixHash (hash n) (hash a)
+  | .lit v       => mixHash 1879 $ hash v
+  | .star        => 7883
+  | .other       => 2411
+  | .arrow       => 17
+  | .proj s i a  =>  mixHash (hash a) $ mixHash (hash s) (hash i)
+
+instance : Hashable Key := ⟨Key.hash⟩
+
+end Key
+
+private def tmpMVarId : MVarId := { name := `_discr_tree_tmp }
+private def tmpStar := mkMVar tmpMVarId
+
+/--
+  Return true iff the argument should be treated as a "wildcard" by the discrimination tree.
+
+  - We ignore proofs because of proof irrelevance. It doesn't make sense to try to
+    index their structure.
+
+  - We ignore instance implicit arguments (e.g., `[Add α]`) because they are "morally" canonical.
+    Moreover, we may have many definitionally equal terms floating around.
+    Example: `Ring.hasAdd Int Int.isRing` and `Int.hasAdd`.
+
+  - We considered ignoring implicit arguments (e.g., `{α : Type}`) since users don't "see" them,
+    and may not even understand why some simplification rule is not firing.
+    However, in type class resolution, we have instance such as `Decidable (@Eq Nat x y)`,
+    where `Nat` is an implicit argument. Thus, we would add the path
+    ```
+    Decidable -> Eq -> * -> * -> * -> [Nat.decEq]
+    ```
+    to the discrimination tree IF we ignored the implicit `Nat` argument.
+    This would be BAD since **ALL** decidable equality instances would be in the same path.
+    So, we index implicit arguments if they are types.
+    This setting seems sensible for simplification theorems such as:
+    ```
+    forall (x y : Unit), (@Eq Unit x y) = true
+    ```
+    If we ignore the implicit argument `Unit`, the `DiscrTree` will say it is a candidate
+    simplification theorem for any equality in our goal.
+
+  Remark: if users have problems with the solution above, we may provide a `noIndexing` annotation,
+  and `ignoreArg` would return true for any term of the form `noIndexing t`.
```

Comment:
```suggestion
  - Proofs are ignored because of proof irrelevance. It doesn't make sense to try to
    index their structure.

  - Instance implicit arguments (e.g., `[Add α]`) are ignored because they are "morally"
    canonical. Moreover, we may have many definitionally equal terms floating around.
    Example: `Ring.hasAdd Int Int.isRing` and `Int.hasAdd`.

  - Implicit arguments (e.g., `{α : Type}`) are not ignored. It may seem that they
    should be, since users don't "see" them,  and may not even understand why
    some simplification rule is not firing.
    However, in type class resolution, there are instances such as `Decidable (@Eq Nat x y)`,
    where `Nat` is an implicit argument. Thus, the path
    ```
    Decidable -> Eq -> * -> * -> * -> [Nat.decEq]
    ```
    would be added to the discrimination tree IF the implicit `Nat` argument were ignored.
    This would be BAD since **ALL** decidable equality instances would be in the same path.
    So, implicit arguments are indexed if they are types.
    This setting seems sensible for simplification theorems such as:
    ```
    forall (x y : Unit), (@Eq Unit x y) = true
    ```
    If we ignore the implicit argument `Unit`, the `DiscrTree` will say it is a candidate
    simplification theorem for any equality in our goal.

  Remark: if users have problems with the solution above, we may provide a `noIndexing` annotation,
  and `ignoreArg` would return true for any term of the form `noIndexing t`.
```

But this docstring seems to include details of `simp` - is it copied from the other discrtree? If so, it'd be good to check that this makes sense still.